### PR TITLE
Fixes #6566 - If the certname does not match the hostname, nil the certname

### PR DESCRIPTION
**Pull request by sodabrew**
Resolves http://projects.theforeman.org/issues/6566 by generalizing the solution from http://projects.theforeman.org/issues/3222

### CODE REVIEW

**Comment by domcleal** [comment]
Could you amend the commit message to "#6566" so Redmine picks up the ticket number?  Thanks.

**Comment by sodabrew** [comment]
[test]

**Comment by GregSutcliffe** [comment]
How does the new cert get to the host? I can see a problem with duplicate hosts showing up in foreman if we're not careful. Consider this:

1) Host is renamed, certname changes along with hostname
2) Host runs puppet, sending old hostname and certname (via the certname param in puppet.conf)
3) Foreman cannot match the cert to an existing host, so creates a new one.

Are we saying it's the users problem to create a new cert matching the new certname? With both values altered at once, Foreman has no way to identify the new host, whereas until now the certname remains the common identifier.

**Comment by GregSutcliffe** [comment]
Oh I'm being a fool, aren't I? This is only during orchestration?

**Comment by sodabrew** [comment]
Yes, this is only at orchestration time. When you put the host into Build mode, Foreman sends a command to the Puppet smart-proxy to invalidate the certificate for `certname`. The question is whether the certname should change for the next rebuild - a valid UUID certname is retained, but a hostname-based certname may need to change to match a new hostname.

File: app/models/host/managed.rb [hunk 1]
```diff
@@ -258,10 +258,9 @@ def handle_ca
...
     return false unless delCertificate
 
-    # If the user has changed use_uuid_for_certificates to false,
-    # then null out the certname. This means we may revoke the hostname
-    # or UUID but will only set autosign for the hostname.
-    if !Setting[:use_uuid_for_certificates] && Foreman.is_uuid?(certname)
+    # If use_uuid_for_certificates is true, reuse the certname UUID value.
+    # If false, then reset the certname if it does not match the hostname.
+    if (Setting[:use_uuid_for_certificates] ? !Foreman.is_uuid?(certname) : certname != hostname)
       logger.info "Removing UUID certificate value #{certname} for host #{name}"
```
Comment (domcleal) [inline app/models/host/managed.rb]: Nitpick, we should probably change this log message to say something like: `"Removing old certificate value #{certname} for host #{name} as it's out of date"`

**Comment by domcleal** [comment]
Looks good to me - are you happy with the change @GregSutcliffe?

File: app/models/host/managed.rb (hunk 1 above)
Comment (sodabrew) [inline app/models/host/managed.rb]: I removed 'UUID' from the comment and rebased.

**Comment by GregSutcliffe** [comment]
I guess my only query, and it's not a bad one, is this: with this in place, is there any value to uuid_certnames? It seems like this is a nicer fix to the problem that uuid certnames was originally implemented for (where hosts want to change hostname because the CR created them with something random).

**Comment by domcleal** [comment]
I think so, because this change is only when doing orchestration or reprovisioning.  uuid_certnames are still useful for hostname changes while a host is running, e.g. a roaming laptop or multi-homed host with an identity crisis.

**Comment by GregSutcliffe** [comment]
That's a very good point. In any case, this is good to go :+1:

**Comment by domcleal** [comment]
Merged as d2823e3 for Foreman 1.6.0.  Thanks @sodabrew!

**Pull request closed without merging**